content/news/2024-01-23-improved-waveforms.md

```diff
@@ -0,0 +1,18 @@
+title: Improved Scrolling Waveforms in Mixxx 2.4.0
+authors: Maarten de Boer
+tags: 2.4, improvements
+status: draft
+comments: yes
+date: 2023-08-15 13:09:47
+
+One of the improvements of Mixxx 2.4 is a rewrite of the scrolling waveforms code. This results in smoother animation at a higher frame rate (60 fps), with less frame-drops and lower CPU load. These improvements are particularly noticeable on macOS, but also apply to Windows and Linux.
+
+The Mixxx scrolling waveforms are drawn with a number of overlapping layers: The beat grid, the looping-, intro- and outro-ranges, the markers, the end-of-track indication, and the actual waveform. Originally these layers were rendered with a combination of Qt (QPainter) and Legacy OpenGL calls, on the now deprecated QGLWidget. Profiling showed that this combination was a performance bottleneck, particularly on macOS. With the rewrite, all these layers are now using Modern OpenGL, with hardware accelerated GLSL shaders. The different waveform types (Simple, Filtered, HSV, RGB and RGB L/R) all come with a GLSL implementation. In addition to the waveforms, the spinny widgets and VU-meters have been rewritten with the same approach.
+
+Visually the new waveforms follow the design of the legacy waveforms, with some minor tweaks (note for example the semi-transparently filled triangles pre- and post track). What has been improved is the layout of the markers layers: Overlapping markers (multiple markers at the same beat grid location) which would previously obfuscate each other, are now automatically placed at an increasing / decreasing vertical position.
```

Comment:
```suggestion
Visually the new waveforms follow the design of the legacy waveforms, with some minor tweaks, note for example the semi-transparently filled triangles pre- and post track. What has been improved is the layout of the markers layers: Overlapping markers (multiple markers at the same beat grid location) which would previously obfuscate each other, are now automatically placed at an increasing / decreasing vertical position.
```
Because every time a bracket is opened, I try to memorize the part before it so that I can resume the previous flow after the bracket has been closed.